### PR TITLE
Feature/tree size limit

### PR DESCRIPTION
**Pull request by cshcomcom**
# Description
- Transactions with greater than 20 tree height are blocked
- Change blockchain.MAX_TX_BYTES to resource.TX_BYTES_LIMIT

Resolve: #311

### CODE REVIEW

File: db/index.js [hunk 1]
```diff
@@ -855,8 +856,14 @@ class DB {
       return false;
     }
     // NOTE(seo): It's not allowed for users to send transactions with auth.fid.
-    return this.executeOperation(
+    const executedResult = this.executeOperation(
         txBody.operation, { addr: tx.address }, txBody.timestamp, tx);
+    const executedStateInfo = this.getStateInfo();
+    if (executedResult && executedStateInfo[StateInfoProperties.TREE_HEIGHT] > TREE_HEIGHT_LIMIT) {
+      return ChainUtil.returnError(910, `Out of tree height limit ` +
+          `(${executedStateInfo[StateInfoProperties.TREE_HEIGHT]} > ${TREE_HEIGHT_LIMIT})`);
```
Comment (platfowner) [inline db/index.js:864]: What about defining a variable for `executedStateInfo[StateInfoProperties.TREE_HEIGHT]`?

File: db/index.js [hunk 2]
```diff
@@ -855,8 +856,14 @@ class DB {
       return false;
     }
     // NOTE(seo): It's not allowed for users to send transactions with auth.fid.
-    return this.executeOperation(
+    const executedResult = this.executeOperation(
```
Comment (platfowner) [inline db/index.js:859]: minor: executedResult sounds like 'the result was executed'. What about executionResult or execResult in short instead?

File: db/index.js [hunk 3]
```diff
@@ -855,8 +856,14 @@ class DB {
...
         txBody.operation, { addr: tx.address }, txBody.timestamp, tx);
+    const executedStateInfo = this.getStateInfo();
```
Comment (platfowner) [inline db/index.js:861]: s/executedStateInfo/stateInfo/?

And getStateInfo() requires input parameter. Could you call `getStateInfo('/')` explicitly?

File: db/index.js [hunk 4]
```diff
@@ -855,8 +856,14 @@ class DB {
...
+    if (executedResult && executedStateInfo[StateInfoProperties.TREE_HEIGHT] > TREE_HEIGHT_LIMIT) {
+      return ChainUtil.returnError(910, `Out of tree height limit ` +
```
Comment (platfowner) [inline db/index.js:863]: Could you add some test cases to `db.test.js`?

File: db/index.js [hunk 5]
```diff
@@ -72,7 +73,7 @@ class DB {
 
   /**
    * Sets state version with its state root.
-   *
```
Comment (platfowner) [inline db/index.js:75]: I'd recommend to turn off the lint rule for these trailing spaces as it seems it doesn't help much for readability or productivity.

File: db/index.js [hunk 6]
```diff
@@ -855,8 +856,14 @@ class DB {
       return false;
     }
     // NOTE(seo): It's not allowed for users to send transactions with auth.fid.
-    return this.executeOperation(
+    const executedResult = this.executeOperation(
         txBody.operation, { addr: tx.address }, txBody.timestamp, tx);
+    const executedStateInfo = this.getStateInfo();
+    if (executedResult && executedStateInfo[StateInfoProperties.TREE_HEIGHT] > TREE_HEIGHT_LIMIT) {
```
Comment (platfowner) [inline db/index.js:862]: The return value of executeOperation() is non-null even with error code, so it needs to explicitly compared with `true` e.g. `result === true`.

File: db/index.js (hunk 2 above)
Comment (cshcomcom) [reply db/index.js:859]: OK. I'll change it to 'executionResult'!

File: db/index.js (hunk 3 above)
Comment (cshcomcom) [reply db/index.js:861]: Looks good! Done

File: db/index.js (hunk 5 above)
Comment (cshcomcom) [reply db/index.js:75]: I agree with your recommendation. But, It needs to set 'no-trailing-spaces' to off. Is it okay?

Comment (cshcomcom) [reply db/index.js:75]: I couldn't find an option to turn off comments only.

File: db/index.js (hunk 6 above)
Comment (cshcomcom) [reply db/index.js:862]: Thanks! In this codes, even if an error is passed, it will pass. I'll change it

File: db/index.js (hunk 5 above)
Comment (platfowner) [reply db/index.js:75]: I see. I think it's just fine to keep it if we don't have any nicer options.

File: db/index.js (hunk 1 above)
Comment (cshcomcom) [reply db/index.js:864]: Done!

File: db/index.js (hunk 4 above)
Comment (cshcomcom) [reply db/index.js:863]: Done!

File: db/index.js (hunk 5 above)
Comment (cshcomcom) [reply db/index.js:75]: Okay!

**Comment by cshcomcom** [comment]
@platfowner All comments are addressed! PTAL

File: unittest/db.test.js [hunk 7]
```diff
@@ -3179,5 +3179,19 @@ describe("Transaction execution", () => {
...
+        operation: {
+          type: 'SET_VALUE',
+          ref: '/test/1/2/3/4/5/6/7/8/9/10/11/12/13/14/15/16/17/18/19/20',
+          value: 'some value',
+        },
+        nonce: -1,
+        timestamp: 1568798344000,
+      };
+      const overHeightTx = Transaction.fromTxBody(overHeightTxBody, node.account.private_key);
+      assert.equal(node.db.executeTransaction(overHeightTx).code, 910);
```
Comment (platfowner) [inline unittest/db.test.js:3194]: what about adding a success case with an input of height 19 to check that the limit is applied preciously?

Comment (cshcomcom) [reply unittest/db.test.js:3206]: Looks good! Done

**Review by platfowner — APPROVED** [review]
LGTM
Thanks!

Could you merge this quickly so that https://github.com/ainblockchain/ain-blockchain/pull/317 is updated?

**Comment by cshcomcom** [comment]
Thanks for reviews!!

**Pull request merged**